### PR TITLE
hotfix/adapt_rails-data-migrations_to_labguru_databases_structure

### PR DESCRIPTION
**Pull request by ShiraUnger**
(cherry picked from commit 4d37fbe375d790d9d8b98983d18a593c70c4f38b)

### CODE REVIEW

**Comment by simonweil** [comment]
First of all, great job, this will reduce deployment time and developer overhead!

A few questions:
1. Is the logic by which the data migrations run the same as the schema migrations? If data migrations run on accounts that didn't get migrations they may fail
2. Won't it be confusing for developers having BioData logic in an external gem?

**Comment by ShiraUnger** [comment]
> First of all, great job, this will reduce deployment time and developer overhead!
> 
> A few questions:
> 
> 1. Is the logic by which the data migrations run the same as the schema migrations? If data migrations run on accounts that didn't get migrations they may fail
> 2. Won't it be confusing for developers having BioData logic in an external gem?

Thanks for the feedback!
1. We check 'active' and 'subscription' and 'churned_at' and maybe more checks when running schema migrations.
I can't see why not run on any existing database, or what is the point of holding a database if it is not aligned with our schema.
I tested my error handling and the process is not crashing if errors occur on one database, and a report of the errors is printed at the end.
I think now that I **should** skip account records with 'db_dropped_at', it can still be handled by this error reporting but it will make the report less readable, and it makes sense to skip these.
2. I hope it will work smoothly enough so they will never have to open it and be confused 😎

File: README.md [hunk 1]
```diff
@@ -1,4 +1,21 @@
...
+# Rails Data Migrations - Labguru's Version
+
+## Labguru Note
+
+At the beggining I used the original gem and thought that we will keep writing the data-migrations ('scripts') on the environment level, 
+when the responsbility on looping through accounts (and the decision on which accounts to run) is on the developer who writes the script hands.
+But then we decided it is better to keep the method similiar to the one we use for database migrations where each migration represent the change to apply on one account/lab.
+I started to override the rake tasks but I had to override each task and also some other files from the gem so I decided it will be better to fork the gem and create our version of it.
+
+In each task I loop through all the accounts but raise an error if the db is missing, so all our databases should be updated even if the account record is not active.
```
Comment (ShiraUnger) [inline README.md:10]: describe error reporting

File: lib/tasks/data_migrations.rake [hunk 2]
```diff
@@ -3,61 +3,144 @@ require 'rake'
...
+      RailsDataMigrations::Migrator.run_migration(
+        direction,
+        RailsDataMigrations::Migrator.migrations_path,
+        version.to_i
+      )
   end
 
   task init_migration: :environment do
-    RailsDataMigrations::LogEntry.create_table
+    errors = []
```
Comment (ShiraUnger) [inline lib/tasks/data_migrations.rake:14]: create table in migrate.

File: README.md [hunk 3]
```diff
@@ -1,4 +1,21 @@
...
+
+To generate a new data migration file -> ```rails generate data_migration fix_owner_id_on_boxes```
+
+To run the migrations -> ```rake data:migrate```
+
+Run a single version -> ```rake data:migrate:up VERSION=20220215065958```
```
Comment (ShiraUnger) [inline README.md:16]: use it only when it can be run again

**Pull request closed without merging**